### PR TITLE
Feature/endpoint new visitors for tables by parameter

### PR DESCRIPTION
**Pull request by KoR82**

### CODE REVIEW

File: backend/app/Http/Controllers/Api/VisitorController.php [hunk 1]
```diff
@@ -92,4 +96,12 @@ public function getVisitorsByParameter (GetTableVisitorsByParameterHttpRequest $
 
         return ApiResponse::success(new TableVisitorsResourseCollection($response->visitors()));
     }
+
+    public function getNewVisitorsForTableByParameter(GetTableVisitorsByParameterHttpRequest $request): ApiResponse
+    {
+        $response = $this->getNewVisitorsByParameterAction->execute(
+            GetVisitorsByParameterRequest::fromRequest($request));
+
+        return ApiResponse::success(new TableVisitorsResourseCollection($response->visitors()));
```
Comment (lenchvolodymyr) [inline backend/app/Http/Controllers/Api/VisitorController.php]: there is general ResourceTable, use it

File: backend/app/Repositories/Contracts/TableVisitorsRepository.php [hunk 2]
```diff
@@ -16,5 +16,5 @@ public function groupByBrowser(int $website_id, string $from, string $to): Colle
 
     public function groupByOperatingSystem(int $website_id, string $from, string $to): Collection;
 
-    public function groupByScreenResolution(int $website_id, string $from, string $to): Collection;
+    public function groupByScreenResolution(int $website_id, string $from, string $to, bool $newVisitors = false): Collection;
```
Comment (lenchvolodymyr) [inline backend/app/Repositories/Contracts/TableVisitorsRepository.php]: never add flags as parameters

Comment (lenchvolodymyr) [reply backend/app/Repositories/Contracts/TableVisitorsRepository.php]: it is the worst practice developers can do. If you add flag as parameter it means that your method will be responsible about two things at least. It is a very rude violation of single responsibility

File: backend/config/app.php [hunk 3]
```diff
@@ -174,6 +174,7 @@
         App\Providers\AuthServiceProvider::class,
         // App\Providers\BroadcastServiceProvider::class,
         App\Providers\EventServiceProvider::class,
+        App\Providers\TelescopeServiceProvider::class,
```
Comment (lenchvolodymyr) [inline backend/config/app.php]: remove this

File: backend/routes/api.php [hunk 4]
```diff
@@ -50,6 +50,8 @@
             Route::get('/new', 'VisitorController@getNewVisitors');
             Route::get('/new/count', 'VisitorController@getNewVisitorsCountForFilterData');
             Route::get('/bounce-rate/total', 'VisitorController@getBounceRate');
+            Route::get('/bounce-rate/total', 'VisitorController@getBounceRate');
```
Comment (lenchvolodymyr) [inline backend/routes/api.php]: why did you duplicate route? be attentive

File: backend/routes/api.php [hunk 5]
```diff
@@ -50,6 +50,8 @@
...
+            Route::get('/by-table/new', 'VisitorController@getNewVisitorsForTableByParameter');
```
Comment (lenchvolodymyr) [inline backend/routes/api.php]: name route new-visitors-table

File: backend/tests/Feature/TableNewVisitorsApiTest.php [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+use Tests\TestCase;
+use Illuminate\Foundation\Testing\WithFaker;
+use Illuminate\Foundation\Testing\RefreshDatabase;
+use Tests\TestDataFactory;
+
+class TableNewVisitorsApiTest extends TestCase
+{
+    use RefreshDatabase;
+
+    const DATE_FROM = '1563148800';
```
Comment (lenchvolodymyr) [inline backend/tests/Feature/TableNewVisitorsApiTest.php]: What is the date "1563148800" ?

File: backend/tests/Feature/TableNewVisitorsApiTest.php [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+        foreach (self::PARAMETERS as $parameter) {
+            $query = [
+                'filter' => [
+                    'start_date' => self::DATE_FROM,
+                    'end_date' => self::DATE_TO
+                ],
+                'parameter' => $parameter
+            ];
+
+            $result = $this->actingAs($this->user)
```
Comment (lenchvolodymyr) [inline backend/tests/Feature/TableNewVisitorsApiTest.php]: you must check not only structure, but also values

Comment (lenchvolodymyr) [reply backend/tests/Feature/TableNewVisitorsApiTest.php]: do not do it in the loop

Comment (lenchvolodymyr) [reply backend/tests/Feature/TableNewVisitorsApiTest.php]: Better create a few visitors and check if the result is correct

File: backend/app/Repositories/EloquentTableVisitorsRepository.php [hunk 8]
```diff
@@ -11,9 +11,16 @@
 
 final class EloquentTableVisitorsRepository implements TableVisitorsRepository
 {
-    public function groupByCity(int $website_id, string $from, string $to): Collection
-    {
+    public function groupByCity(
```
Comment (lenchvolodymyr) [inline backend/app/Repositories/EloquentTableVisitorsRepository.php]: Create another repository. New visitors and visitors are different entities. And moreover you are passing flag as parameter

File: backend/app/Actions/Visitors/GetNewVisitorsByParameterAction.php [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+                        Auth::user()->website->id,
+                        $request->startDate(),
+                        $request->endDate(),
+                        true
+                    );
+                break;
+            default:
+                throw new InvalidArgumentException(sprintf('The parameter "%s" is not valid.', $request->parameter()));
+        }
+
```
Comment (lenchvolodymyr) [inline backend/app/Actions/Visitors/GetNewVisitorsByParameterAction.php]: what is this space?

File: backend/app/Http/Controllers/Api/VisitorController.php (hunk 1 above)
Comment (KoR82) [reply backend/app/Http/Controllers/Api/VisitorController.php]: Fixed

File: backend/app/Repositories/Contracts/TableVisitorsRepository.php (hunk 2 above)
Comment (KoR82) [reply backend/app/Repositories/Contracts/TableVisitorsRepository.php]: Fixed

File: backend/config/app.php (hunk 3 above)
Comment (KoR82) [reply backend/config/app.php]: Fixed

File: backend/routes/api.php (hunk 4 above)
Comment (KoR82) [reply backend/routes/api.php]: Fixed

File: backend/routes/api.php (hunk 5 above)
Comment (KoR82) [reply backend/routes/api.php]: Fixed

File: backend/tests/Feature/TableNewVisitorsApiTest.php (hunk 6 above)
Comment (KoR82) [reply backend/tests/Feature/TableNewVisitorsApiTest.php]: Fixed

File: backend/app/Repositories/EloquentTableVisitorsRepository.php (hunk 8 above)
Comment (KoR82) [reply backend/app/Repositories/EloquentTableVisitorsRepository.php]: Fixed

File: backend/app/Actions/Visitors/GetNewVisitorsByParameterAction.php (hunk 9 above)
Comment (KoR82) [reply backend/app/Actions/Visitors/GetNewVisitorsByParameterAction.php]: Fixed

**Pull request merged**